File: DesarrolloSRI/FacturacionElectronica/DaxFelec/WebServiseSRI/WebService/EnviarComprobanteSri.vb
Imports System.Xml
Imports System.Text

Public Class EnviarComprobanteSri
    Dim pathError As String

    Public Function sendComprobanteSRI(ByVal fileName As String, ByVal clave As String, ByVal pathAutorizado As String, ByVal pathNoAutorizados As String) As Boolean
        Dim ws As New WebService.WebServicesSRI
        Dim autorizado As Boolean = False
        Dim errores As New List(Of String)
        Dim eror As String
        '--------------------------------------------------------------------------------------------
        pathError = pathNoAutorizados.Replace("XML", "err")
        pathError = pathError.Replace("xml", "err")
        Try
            If ws.sendComprobante(fileName) = True Then
                If ws.AutorizacionComprobante.IsAutorizado = True Then
                    autorizado = True
                    Dim aut As WebService.Autorizacion = ws.AutorizacionComprobante.Autorizacion
                    '--------------------------------------------------------------------------------
                    Dim util As New adcFeutil.Feutilidad()
                    util.ActualizarDocumentoAdcom("Autorizado", "", "", 0, clave, aut.NumeroAutorizacion, 0, aut.FechaAutorizacion)
                    'System.IO.File.Copy(fileName, pathAutorizado, True)
                    GuardarArchivoAutorizado(aut, pathAutorizado)
                    'MsgBox(String.Format("Clave de Acceso:{1}{0}Tipo Emision:{2}{0}Numero Autorizacion:{3}{0}Fecha Autorizacion:{4}{0}", vbCrLf, ws.ClaveAcceso, ws.TipoEmision, aut.NumeroAutorizacion, aut.FechaAutorizacion))
                    '--------------------------------------------------------------------------------                    
                    aut = Nothing
                ElseIf ws.AutorizacionComprobante.NumeroComprobantes > 0 Then
                    System.IO.File.Copy(fileName, pathNoAutorizados, True)
                    For Each mensaje As WebService.Mensaje In ws.AutorizacionComprobante.Autorizacion.Mensajes
                        eror = mensaje.Mensaje + vbCr + mensaje.InformacionAdicional
                        errores.Add(eror)
                    Next
                End If
            ElseIf ws.RecepcionComprobante.NumeroComprobantes > 0 Then
                System.IO.File.Copy(fileName, pathNoAutorizados, True)
                For Each mensaje As WebService.Mensaje In ws.RecepcionComprobante.Comprobante.Mensajes
                    eror = mensaje.Mensaje + vbCr + mensaje.InformacionAdicional
                    errores.Add(eror)
                Next
            ElseIf ws.IsOnLine = False Then
                System.IO.File.Copy(fileName, pathNoAutorizados, True)
                '                grabarMensajes("No existe conexión con los servicios Web del SRI " + Now.ToLongDateString)
                eror = "ERROR: No existe conexión con los servicios Web del SRI " + Now.ToLongDateString
                errores.Add(eror)
            End If
            '----------------------------------------------------------------------------------------
        Catch ex As Exception
            'MsgBox(ex.Message, MsgBoxStyle.Critical)
            eror = ex.Message + " " + Now.ToLongDateString
            errores.Add(eror)
        End Try

        '--------------------------------------------------------------------------------------------

        If autorizado = False Then
            If errores.Count > 0 Then grabarMensajes(errores)
        Else
            Try
                'System.IO.File.Delete(fileName)
            Catch ee As Exception
                errores.Add("Excepcion eliminando firmado: " & ee.Message)
                grabarMensajes(errores)
            End Try
        End If
        ws = Nothing
        Return autorizado
    End Function

    Private Sub grabarMensajes(mensajes As List(Of String))
        Dim file As New System.IO.StreamWriter(pathError)

        For Each Mensaje As String In mensajes
            file.WriteLine(Mensaje)
        Next

        file.Flush()
        file.Close()
        file = Nothing

    End Sub

    Private Sub GuardarArchivoAutorizado(ByVal comprobante As Autorizacion, ByVal outpahFile As String)
        Dim info As New System.IO.FileInfo(outpahFile)
        '---------------------------------------------------------------------------------------------
        'verifico si existe el directorio
        '---------------------------------------------------------------------------------------------
        If info.Directory.Exists = True Then
            Dim writer As New XmlTextWriter(outpahFile, New UTF8Encoding(True))
            '-----------------------------------------------------------------------------------------
            writer.WriteProcessingInstruction("xml", "version=""1.0"" encoding=""utf-8""")
            '-----------------------------------------------------------------------------------------
            writer.WriteStartElement("autorizacion")
            '-----------------------------------------------------------------------------------------
            writer.WriteElementString("estado", comprobante.Estado)
            '-----------------------------------------------------------------------------------------
            writer.WriteElementString("numeroAutorizacion", comprobante.NumeroAutorizacion)
            '-----------------------------------------------------------------------------------------
            writer.WriteStartElement("fechaAutorizacion")
            writer.WriteAttributeString("class", "fechaAutorizacion")
            writer.WriteString(comprobante.FechaAutorizacion)
            writer.WriteEndElement()
            '-----------------------------------------------------------------------------------------
            writer.WriteStartElement("comprobante")
            writer.WriteCData(comprobante.Comprobante)
            writer.WriteEndElement()
            '-----------------------------------------------------------------------------------------
            writer.WriteEndElement()
            '-----------------------------------------------------------------------------------------
            writer.Flush()
            writer.Close()
            '-----------------------------------------------------------------------------------------
            writer = Nothing
        End If
        '---------------------------------------------------------------------------------------------
        info = Nothing
    End Sub
End Class

File: DesarrolloSRI/FacturacionElectronica/enviarDocumentoSri/EnviarComprobanteSri.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using WebServiceSri;

namespace DaxDocElectronicos
{
    //    public class EnviarComprobanteSri
    //    {
    //        string pathError ="";
    //        public Boolean sendComprobanteSRI(string fileName, string clave, string pathAutorizado, string pathNoAutorizados,Boolean OnnLine,string ambiente, string strConxAdcom)
    //        {            
    //            WebServiceSri.WebService ws = new WebServiceSri.WebService();
    //            WebServiceSri.RespuestaSRYType respSri = WebServiceSri.RespuestaSRYType.NOAUTORIZADO;
    //            ws.IsOFFLine = !OnnLine;

    //            bool autorizado = false;
    //            List<string> errores = new List<string>();
    //            string error = "";
    //            //--------------------------------------------------------------------------------------------
    //            pathError = pathNoAutorizados.Replace("XML", "err");
    //            pathError = pathError.Replace("xml", "err");

    //            //try
    //            {
    //                respSri = ws.SendComprobante(fileName, pathAutorizado);
    //                if (respSri == WebServiceSri.RespuestaSRYType.AUTORIZADO)
    //                {
    //                    autorizado = true;
    //                    Feutilidad util = new Feutilidad();
    //                    util.ActualizarDocumentoAdcom("Autorizado", "", "", 0, clave, ws.Comprobante.NumeroAutorizacion, 0, ws.Comprobante.FechaAutorizacion,Convert.ToInt16(ambiente),strConxAdcom);
    //                }
    //                else if (ws.Comprobante.Mensajes.Count > 0)
    //                {
    //                    System.IO.File.Copy(fileName, pathNoAutorizados, true);
    //                    System.IO.File.Delete(pathAutorizado);
    //                    error = ws.Comprobante.getMensajes();
    //                }
    //            }
    ////            catch (Exception ex)
    //            {
    //  //              error = ex.Message + " " + DateTime.Now.ToLongDateString();
    //            }

    //            if (error.Length > 0) { errores.Add(error); grabarMensajes(errores);}
    //            ws = null;
    //            return autorizado;
    //        }
    //        private void grabarMensajes( List<String> mensajes)
    //            {
    //                System.IO.StreamWriter file = new System.IO.StreamWriter(pathError);

    //                foreach (string Mensaje in mensajes)
    //                {
    //                    file.WriteLine(Mensaje);
    //                }
    //                file.Flush();
    //                file.Close();
    //                file.Dispose();
    //            }
    //    }
    public class EnviarComprobanteSri
    {
        private string pathError = "";

        public bool sendComprobanteSRI(string fileName,string clave,string pathAutorizado,string pathNoAutorizados,bool OnnLine,string ambiente,string strConxAdcom)
        {
            WebService webService = new WebService();
            webService.IsOFFLine = !OnnLine;
            bool flag = false;
            List<string> mensajes = new List<string>();
            string str = "";
            this.pathError = pathNoAutorizados.Replace("XML", "err");
            this.pathError = this.pathError.Replace("xml", "err");
            try
            {
                if (webService.SendComprobante(fileName, pathAutorizado) == RespuestaSRYType.AUTORIZADO)
                {
                    flag = true;
                    new Feutilidad().ActualizarDocumentoAdcom("Autorizado", "", "", 0.0, clave, webService.Comprobante.NumeroAutorizacion, 0, webService.Comprobante.FechaAutorizacion, (int)Convert.ToInt16(ambiente), strConxAdcom);
                }
                else if (webService.Comprobante.Mensajes.Count > 0)
                {
                    File.Copy(fileName, pathNoAutorizados, true);
                    File.Delete(pathAutorizado);
                    str = webService.Comprobante.getMensajes();
                }
            }
            catch (Exception ex)
            {
                str = ex.Message + " " + DateTime.Now.ToLongDateString();
            }
            if (str.Length > 0)
            {
                mensajes.Add(str);
                this.grabarMensajes(mensajes);
            }
            return flag;
        }

        private void grabarMensajes(List<string> mensajes)
        {
            StreamWriter streamWriter = new StreamWriter(this.pathError);
            foreach (string mensaje in mensajes)
                streamWriter.WriteLine(mensaje);
            streamWriter.Flush();
            streamWriter.Close();
            streamWriter.Dispose();
        }
    }



}
